In [2]:
import pandas as pd

In [3]:
# Load datasets
contract_df = pd.read_csv('/datasets/contract.csv')
personal_df = pd.read_csv('/datasets/personal.csv')
internet_df = pd.read_csv('/datasets/internet.csv')
phone_df = pd.read_csv('/datasets/phone.csv')

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> See later comment, you are pulling data from the wrong place, these should be from the '/datasets/final_provider/' folder.
<a class="tocSkip"></


In [4]:
# Get a general overview of the data
contract_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   BeginDate         7043 non-null   object 
 2   EndDate           7043 non-null   object 
 3   Type              7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
dtypes: float64(1), object(7)
memory usage: 440.3+ KB


In [5]:
# View the data
contract_df

customerID   BeginDate              EndDate            Type  \
0     7590-VHVEG  2020-01-01                   No  Month-to-month   
1     5575-GNVDE  2017-04-01                   No        One year   
2     3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3     7795-CFOCW  2016-05-01                   No        One year   
4     9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   
...          ...         ...                  ...             ...   
7038  6840-RESVB  2018-02-01                   No        One year   
7039  2234-XADUH  2014-02-01                   No        One year   
7040  4801-JZAZL  2019-03-01                   No  Month-to-month   
7041  8361-LTMKD  2019-07-01  2019-11-01 00:00:00  Month-to-month   
7042  3186-AJIEK  2014-08-01                   No        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
0                 Yes           Electronic check           29.85        29.85  
1                  No               Mailed check           56.95       1889.5  
2                 Yes               Mailed check           53.85       108.15  
3                  No  Bank transfer (automatic)           42.30      1840.75  
4                 Yes           Electronic check           70.70       151.65  
...               ...                        ...             ...          ...  
7038              Yes               Mailed check           84.80       1990.5  
7039              Yes    Credit card (automatic)          103.20       7362.9  
7040              Yes           Electronic check           29.60       346.45  
7041              Yes               Mailed check           74.40        306.6  
7042              Yes  Bank transfer (automatic)          105.65       6844.5  

[7043 rows x 8 columns]

In [16]:
# Find the percentage of current customers
len(contract_df[contract_df['EndDate'] == 'No'])/ len(contract_df['EndDate'])

0.7346301292063041

#### **Notes About Contract Dataset**
In the contract dataset, we have 7043 customers accounted for. About 73% of customers have not turned which indicates we have an imbalanced class. We also have no missing data.

**Preprocessing Tasks**
- Column names and string values need to be standardized to a lowercase format, remove whitespace and use of underscore to separate words.
- Change 'TotalCharges' column into a float64
- Convert 'BeginDate' to a datetime format

**Feature Engineering Tasks**
- 'Type', 'PaperlessBilling' and 'PaymentMethod' will need to be converted categorically using label encoding (paperless billing), one hot encoding (payment methode) and ordinal encoding (type)
- 'EndDate' is a binary target and should be defined as 1 = churn and 0 = no_churn in a new column called 'churn'
- Check seasonalty for BeginDate to identify if they only signed up for the deal and add 'contract_age' column
- Perform class balancing on the training set after the split and before model training

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> Nice job writing out your to-dos.
<a class="tocSkip"></a>

In [7]:
personal_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     7043 non-null   object
 1   gender         7043 non-null   object
 2   SeniorCitizen  7043 non-null   int64 
 3   Partner        7043 non-null   object
 4   Dependents     7043 non-null   object
dtypes: int64(1), object(4)
memory usage: 275.2+ KB


In [8]:
personal_df

customerID  gender  SeniorCitizen Partner Dependents
0     7590-VHVEG  Female              0     Yes         No
1     5575-GNVDE    Male              0      No         No
2     3668-QPYBK    Male              0      No         No
3     7795-CFOCW    Male              0      No         No
4     9237-HQITU  Female              0      No         No
...          ...     ...            ...     ...        ...
7038  6840-RESVB    Male              0     Yes        Yes
7039  2234-XADUH  Female              0     Yes        Yes
7040  4801-JZAZL  Female              0     Yes        Yes
7041  8361-LTMKD    Male              1     Yes         No
7042  3186-AJIEK    Male              0      No         No

[7043 rows x 5 columns]

#### **Notes About Personal Dataset**
In the personal dataset, we have 7043 customers accounted for. We also have no missing data.

**Preprocessing Tasks**
- Column names need to be standardized to a lowercase format, remove whitespace and use of underscore to separate words.
- String values need to be standardized to a lowercase format


**Feature Engineering Tasks**
- 'gender', 'Partner' and 'Dependents' will need to be converted categorically using label encoding
- After preprocessing, merge with contract_df on 'customerID'

In [9]:
internet_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149396 entries, 0 to 149395
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Unnamed: 0    149396 non-null  int64  
 1   id            149396 non-null  object 
 2   mb_used       149396 non-null  float64
 3   session_date  149396 non-null  object 
 4   user_id       149396 non-null  int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 5.7+ MB


In [10]:
internet_df

Unnamed: 0        id  mb_used session_date  user_id
0                0    1000_0   112.95   2018-11-25     1000
1                1    1000_1  1052.81   2018-09-07     1000
2                2    1000_2  1197.26   2018-06-25     1000
3                3    1000_3   550.27   2018-08-22     1000
4                4    1000_4   302.56   2018-09-24     1000
...            ...       ...      ...          ...      ...
149391      149391  1499_152   318.90   2018-10-03     1499
149392      149392  1499_153   490.13   2018-12-14     1499
149393      149393  1499_154     0.00   2018-10-27     1499
149394      149394  1499_155  1246.32   2018-11-26     1499
149395      149395  1499_156   544.37   2018-10-26     1499

[149396 rows x 5 columns]

In [11]:
internet_df['user_id'].value_counts()

1496    834
1227    812
1204    789
1240    764
1251    759
       ... 
1151     35
1467     27
1343     26
1192     22
1399      8
Name: user_id, Length: 497, dtype: int64

#### **Notes About Internet Dataset**
In the internet dataset, we have 497 out of 7043 customers accounted for using internet. We also have no missing data. Immediately I noticed there are no columns titled 'customerID' like the other datasets. 

**Preprocessing Tasks**
- The 'Unnamed: 0' column appears to be a duplicate of the index so it can be removed since it doesn't have values of prominence.
- Convert 'session_date' to a datetime

**Feature Engineering Tasks**
- If the 'user_id' number aligns with the first 4 characters of the contract_df 'customerID' column, we will have to create a 'user_id' column in the other datasets and merge them on user_id
- Get monthly mb_usage per customer since most plans are determined by monthly usage

**Clarifying Questions**
- Does the first 4 digits of the 'customerID' column align with the 4 digits of the 'user_id' column in the internet_df?
- Should I create an additional binary column (e.g., has_internet) to indicate whether the customer has any internet service?

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> Answers to quesions:<br>
1. As previously stated you need to pull from the '/datasets/final_provider/' folder. The internet dataset will look different after this. <br>
    2. You are welcome to, but you will know who has internet based on if there are NA's after merging.
<a class="tocSkip">

In [12]:
phone_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6361 entries, 0 to 6360
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     6361 non-null   object
 1   MultipleLines  6361 non-null   object
dtypes: object(2)
memory usage: 99.5+ KB


In [13]:
phone_df

customerID MultipleLines
0     5575-GNVDE            No
1     3668-QPYBK            No
2     9237-HQITU            No
3     9305-CDSKC           Yes
4     1452-KIOVK           Yes
...          ...           ...
6356  2569-WGERO            No
6357  6840-RESVB           Yes
6358  2234-XADUH           Yes
6359  8361-LTMKD           Yes
6360  3186-AJIEK            No

[6361 rows x 2 columns]

In [14]:
# Check for duplicates
phone_df['customerID'].duplicated().sum()

0

In [15]:
# Find customerIDs in contract_df that are not in phone_df
missing_phone_customers = contract_df[~contract_df['customerID'].isin(phone_df['customerID'])]

# Display only the customerIDs
missing_ids = missing_phone_customers['customerID']
print(missing_ids)

0       7590-VHVEG
3       7795-CFOCW
7       6713-OKOMC
20      8779-QRDMV
27      8665-UTDHZ
           ...    
7003    4501-VCPFK
7007    2274-XUATA
7029    2235-DWLJU
7036    7750-EYXWZ
7040    4801-JZAZL
Name: customerID, Length: 682, dtype: object


#### **Notes About Internet Dataset**
In the phone dataset, we have 6361 out of 7043 customers accounted for having multiple phone lines or not. There are no columns to distinguish if the customer has phone service or not. We also have no missing data. Immediately I noticed there are no columns titled 'customerID' like the other datasets. 

**Preprocessing Tasks**
- Column names need to be standardized to lowercase and add underscores

**Feature Engineering Tasks**
- the 'MultipleLines' column needs a categorical encoding like the others. I will use one-hot encoding for this column.

**Clarifying Questions**
- What does 'MultipleLines' actually represent? Does it refer to multiple phone lines under one account, or multiple users on one line?

- How should the 682 customers without multiple lines be treated — included with missing values or excluded from the analysis?

#### **Proposed Work Plan**
The goal of this project is to develop a model that will determine if a customer will leave. These are the steps that I will take to accomplish this task;

1. Download the data
2. Explore the data to determine how to handle it in preprocessing
3. Perform preprocessing on the data to include;
    - Merge all datasets
    - Standardize the format to lowercase with underscores on columns and values
    - Convert datatypes to most appropriate type for feature engineering
      
5. Perform feature engineering;
     - Before merging, create a user_id column for all datasets based off of the first 4 characters in the customerID then merge on user_id.
     - Create new features:
         - 'Churn' with 1 = churn and 0 = not churn
         - 'Tenure' from 'BeginDate'
         - 'Long_Term_Contract' = contract != 'month-to-month'
     - Encode categorical variables:
         - Use Label encoding for binary features: 'gender', 'Partner', 'Dependents', 'PaperlessBilling'
         - Ordinal encoding for ordered categories: 'Type'
         - One-hot encoding: 'PaymentMethod', 'MultipleLines', 'InternetService'

8. Fill in missing values that occurred due to the merge with 'unknown'
9. Perform EDA (Exploratory Data Analysis)
    - Visualize churn distribution
    - Analyze relationships between features and churn
    - Identify imbalance and correlation
11. Model training and evaluation
    - Split data: training, validation, and test sets (e.g., 70/15/15).
    - Handle class imbalance:
        - Use class_weight='balanced' or apply SMOTE on the training set
    - Train and compare multiple models:
        - Logistic Regression (baseline)
        - Random Forest
        - XGBoost or LightGBM (for performance)
    - Tune hyperparameters using GridSearchCV or RandomizedSearchCV
    - Evaluate with:
        - AUC-ROC (primary)
        - Accuracy, precision, recall (secondary)
12. Write a conclusion

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> This looks like a good plan here! Nice job and best of luck. 
<a class="tocSkip"></a>